homeassistant/components/stream/core.py

```diff
@@ -386,17 +386,27 @@ async def handle(
         raise NotImplementedError()
 
 
-TRANSFORM_IMAGE_FUNCTION = (
-    lambda image: image,  # Unused
-    lambda image: image,  # No transform
-    lambda image: np.fliplr(image).copy(),  # Mirror
-    lambda image: np.rot90(image, 2).copy(),  # Rotate 180
-    lambda image: np.flipud(image).copy(),  # Flip
-    lambda image: np.flipud(np.rot90(image)).copy(),  # Rotate left and flip
-    lambda image: np.rot90(image).copy(),  # Rotate left
-    lambda image: np.flipud(np.rot90(image, -1)).copy(),  # Rotate right and flip
-    lambda image: np.rot90(image, -1).copy(),  # Rotate right
-)
+def _transform_image(image: np.ndarray, orientation: int) -> np.ndarray:
+    """Transform the image using numpy."""
+
+    # Keep import here so that we can import stream integration without installing reqs
+    import numpy as np  # pylint: disable=import-outside-toplevel
+
+    if orientation == 2:  # Mirror
+        return np.fliplr(image).copy()
+    if orientation == 3:  # Rotate 180
+        return np.rot90(image, 2).copy()
+    if orientation == 4:  # Flip
+        return np.flipud(image).copy()
+    if orientation == 5:  # Rotate left and flip
+        return np.flipud(np.rot90(image)).copy()
+    if orientation == 6:  # Rotate left
+        return np.rot90(image).copy()
+    if orientation == 7:  # Rotate right and flip
+        return np.flipud(np.rot90(image, -1)).copy()
+    if orientation == 8:  # Rotate right
+        return np.rot90(image, -1).copy()
+    return image  # Transforms 0 and 1 (and others not yet defined) are identity
```

Comment:
```suggestion
    return np.rot90(image, -1).copy()  # Rotate right
```

Reply:
See above

Reply:
I'd rather not have an out-of-range value mean "rotate right" 😁

Reply:
If you don't apply, then you end up with a gap in code coverage.

Reply:
It won't get out of range. It is initialized to 1 and only gets changed via websocket which restricts it to 1 through 8.

Reply:
https://github.com/home-assistant/core/blob/2cb58b818955a3291e25da62f1b06c2b8f2ce248/homeassistant/components/camera/__init__.py#L871

Reply:
So long as that's the only way this will ever be called. I don't see the issue in falling through to returning the original – the other good option in my books would be to `raise` on an invalid orientation.

Reply:
That is the only way it's currently called. If code was added that calls it a different way then reviewers should be aware of the behavior - perhaps add a comment?
Even if an unexpected value is encountered, I don't think returning a different rotation on an unexpected value is really any better/worse than returning the original (if we are being anal then we should be raising regardless).
I don't see a good reason to complicate the code here.

Reply:
> That is the only way it's currently called.

My point is that at some point someone might refactor it elsewhere for some other use and start using it without knowing you'd need to validate the `orientation` argument first, and at that point it becomes "oh that's the weird function that turns your picture sideways if you call it wrong".

Anyway, pushed a change that adds internal validation (in the early-return branch) and gets rid of the last if...

Reply:
Why is it weird to turn it sideways if it's called wrong? I would argue that's probably even better than passing it unchanged as at least you know something is wrong.
This is an internal function that's used only in stream. If someone tries to use it elsewhere they should read it and figure out how it works.
Perhaps the orientation should be typed as an enum but that's beyond the scope of this PR.
I would really prefer it if you could change it as in my original suggestion.

Reply:
> Why is it weird to turn it sideways if it's called wrong? I would argue that's probably even better than passing it unchanged as at least you know something is wrong.

I would not want to be the future _someone_ debugging why on earth some images are being turned sideways for some reason, and to maybe eventually find out the reason is somehow something is passing 9 to this function.

The [principle of least astonishment](https://en.wikipedia.org/wiki/Principle_of_least_astonishment) applies here, too.

If you really strongly do feel that this function should rotate images 90 degrees right when miscalled, I can begrudgingly do that, but that's also _not_ what the old implementation did (which was to raise a `KeyError`).

Reply:
I've told you what I think already. This is an internally used function that only gets passed ints from 1 to 8. You've added extra code to account for other values to satisfy some hypothetical future where another caller uses it wrongly.
I think if we're very concerned about that, just add a comment clarifying that it should only be called with 1 through 8 and will default to 8.